src/java/org/apache/cassandra/db/commitlog/CommitLogDescriptor.java

```diff
@@ -43,38 +51,56 @@
     public static final int VERSION_12 = 2;
     public static final int VERSION_20 = 3;
     public static final int VERSION_21 = 4;
+    public static final int VERSION_22 = 5;
     /**
      * Increment this number if there is a changes in the commit log disc layout or MessagingVersion changes.
      * Note: make sure to handle {@link #getMessagingVersion()}
      */
-    public static final int current_version = VERSION_21;
-
-    // [version, id, checksum]
-    static final int HEADER_SIZE = 4 + 8 + 4;
+    public static final int current_version = VERSION_22;
 
     final int version;
     public final long id;
+    public final ParametrizedClass compression;
 
-    public CommitLogDescriptor(int version, long id)
+    public CommitLogDescriptor(int version, long id, ParametrizedClass compression)
     {
         this.version = version;
         this.id = id;
+        this.compression = compression;
     }
 
     public CommitLogDescriptor(long id)
     {
-        this(current_version, id);
+        this(current_version, id, DatabaseDescriptor.getCommitLogCompression());
     }
 
-    static void writeHeader(ByteBuffer out, CommitLogDescriptor descriptor)
+    public static void writeHeader(ByteBuffer out, CommitLogDescriptor descriptor)
     {
-        out.putInt(0, descriptor.version);
-        out.putLong(4, descriptor.id);
         PureJavaCrc32 crc = new PureJavaCrc32();
+        out.putInt(descriptor.version);
         crc.updateInt(descriptor.version);
+        out.putLong(descriptor.id);
         crc.updateInt((int) (descriptor.id & 0xFFFFFFFFL));
         crc.updateInt((int) (descriptor.id >>> 32));
-        out.putInt(12, crc.getCrc());
+        String compressionString = constructCompressionString(descriptor.compression);
+        if (descriptor.version >= VERSION_22) {
+            byte[] compressionBytes = compressionString.getBytes(StandardCharsets.UTF_8);
+            out.putShort((short) compressionBytes.length);
+            crc.updateInt(compressionBytes.length);
+            out.put(compressionBytes);
+            crc.update(compressionBytes, 0, compressionBytes.length);
+        }
```

Comment:
It might be more future proof to put a JSON doc at the header and nest the compression settings in that. Adding attributes in future versions will be a little cleaner since the boiler plate for a variable length field serializing and checksumming will not have to be repeated at both ends.